{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Complilation of auxiliary data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import re, os, sys#, math, pickle, subprocess, resource, random\n",
    "from pathlib import Path\n",
    "import numpy\n",
    "import pandas\n",
    "from pandas import DataFrame, Series\n",
    "\n",
    "numpy.random.seed(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "root_dir = Path(os.environ['HOME'], 'simons/faststorage/people/kmt')\n",
    "#meta_data_dir = Path(os.environ['HOME'], 'simons/faststorage/data/metadata')\n",
    "#steps_dir = root_dir / 'steps'\n",
    "#argweaver_dir = steps_dir / 'argweaver/output'\n",
    "results_dir = root_dir / 'results'\n",
    "#figures_dir = root_dir / 'figures'\n",
    "data_dir = root_dir / 'data'\n",
    "#pi_dir = steps_dir / 'pi_stores'\n",
    "#dist_dir = steps_dir / 'dist_stores'\n",
    "#male_x_haploid_dir = steps_dir / 'male_x_haploids'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local code in the scripts dir on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scripts_dir = root_dir / 'scripts'\n",
    "if str(scripts_dir) not in sys.path:\n",
    "    sys.path.append(str(scripts_dir))\n",
    "\n",
    "import simons_meta_data\n",
    "import hg19_chrom_sizes\n",
    "\n",
    "from GenomicWindows import window\n",
    "import GenomicIntervals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convenience functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def silent_nanmean(x):\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        return numpy.nanmean(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# Make inline plots vector graphics instead of raster graphics\n",
    "from IPython.display import set_matplotlib_formats\n",
    "set_matplotlib_formats('pdf', 'svg')\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.patches import Rectangle\n",
    "from mpl_toolkits.basemap import Basemap\n",
    "#matplotlib.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "\n",
    "import mpld3\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_style(\"whitegrid\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "# lowess for plotting\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "from scipy.interpolate import UnivariateSpline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# silence deprecation warnings (lots from seaborn)\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning) \n",
    "warnings.filterwarnings(\"ignore\", category=numpy.VisibleDeprecationWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load genome annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ampliconic regions and Shriram's regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ampliconic_regions = (pandas.read_table(str(data_dir / 'coordinates_hg18_hg19_hg38_Amplicons_Gap.txt'))\n",
    "                      .assign(start=lambda df: df.hg19_start, \n",
    "                              end=lambda df: df.hg19_end,\n",
    "                              chrom='X')\n",
    "                      .loc[:, ['chrom', 'start', 'end']]\n",
    "                     )\n",
    "ampliconic_regions.head()\n",
    "\n",
    "# ampliconic_regions = (pandas.read_table(str(data_dir / 'AmpliconicRegions_hg19.csv'),\n",
    "#                                        header=0,\n",
    "#                                         names=['chrom', 'start', 'end', 'region', 'midpoint',\n",
    "#                                                         'size', 'alignment', 'pnilsr']\n",
    "#                                        )\n",
    "#                       .loc[:, ['chrom', 'start', 'end']]\n",
    "#                       .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "#                      )\n",
    "\n",
    "ampliconic_regions.to_hdf(results_dir / 'ampliconic_regions.hdf', 'df', mode='w')\n",
    "\n",
    "no_neanderthal_asn = (pandas.read_table(str(data_dir / 'Neanderthal_ASN.csv'), sep=' ', \n",
    "                                       header=0, names=['chrom', 'start', 'end'])\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "no_neanderthal_asn.to_hdf(results_dir / 'no_neanderthal_asn.hdf', 'df', mode='w')\n",
    "\n",
    "\n",
    "no_neanderthal_eur = (pandas.read_table(str(data_dir / 'Neanderthal_EUR.csv'), sep=' ',\n",
    "                                       header=0, names=['chrom', 'start', 'end'])\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "no_neanderthal_eur.to_hdf(results_dir / 'no_neanderthal_eur.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RefSeq genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "refseq_genes = (pandas.read_table(str(data_dir / 'refseq_genes_hg19_table.txt')))\n",
    "refseq_genes.chrom = [x.replace('chr', '') for x in refseq_genes.chrom]\n",
    "refseq_genes.rename(columns={\"txStart\": \"start\", \"txEnd\": 'end'}, inplace=True)\n",
    "refseq_genes.sort_values(by=['chrom', 'start', 'end'], inplace=True)\n",
    "\n",
    "refseq_genes.to_hdf(results_dir / 'refseq_genes.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Density of transcribed Refseq gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "window_size = 100000\n",
    "@window(size=window_size, fill={'gene_density': 0})\n",
    "def gene_density(df):\n",
    "    df = GenomicIntervals.interval_collapse(df)\n",
    "\n",
    "    return {'gene_density': sum(df.end - df.start) / window_size}\n",
    "\n",
    "gene_density_df = refseq_genes.groupby('chrom').apply(gene_density).reset_index(level=['chrom'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Density of refseq exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def get_exons(df):\n",
    "    exon_starts = list()\n",
    "    exon_ends = list()\n",
    "    for row in df.itertuples():\n",
    "        chrom = row.chrom\n",
    "        exon_starts.extend(list(map(int, row.exonStarts.split(',')[:-1])))\n",
    "        exon_ends.extend(list(map(int, row.exonEnds.split(',')[:-1])))\n",
    "    all_exons = DataFrame({'chrom': chrom, 'start': exon_starts, 'end': exon_ends})\n",
    "    all_exons.sort_values(['start', 'end'], inplace=True)\n",
    "    collapsed_exons = GenomicIntervals.interval_collapse(all_exons)\n",
    "    return collapsed_exons\n",
    "\n",
    "# data frame with exon start end coordinates\n",
    "refseq_exons = refseq_genes.groupby('chrom').apply(get_exons).reset_index(drop=True)\n",
    "\n",
    "window_size = 100000\n",
    "@window(size=window_size)\n",
    "def exon_density(df):\n",
    "    return {'exon_density': sum(df.end - df.start) / window_size}\n",
    "\n",
    "# windowed exon density\n",
    "exon_density_df = refseq_exons.groupby('chrom').apply(exon_density).reset_index(level=['chrom'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge both into one data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "refseq_density = pandas.merge(exon_density_df, gene_density_df, on=['chrom', 'start', 'end'])\n",
    "\n",
    "refseq_density.to_hdf(results_dir / 'refseq_density.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mirBase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GFF2 in hg19 ftp://mirbase.org/pub/mirbase/20/genomes/hsa.gff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>method</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17368</td>\n",
       "      <td>17434</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0022705\"; ID=\"hsa-mir-6859-1\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30365</td>\n",
       "      <td>30501</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0006363\"; ID=\"hsa-mir-1302-2\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>567704</td>\n",
       "      <td>567791</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0022558\"; ID=\"hsa-mir-6723\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>1102483</td>\n",
       "      <td>1102576</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0000342\"; ID=\"hsa-mir-200b\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>1103242</td>\n",
       "      <td>1103330</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0000737\"; ID=\"hsa-mir-200a\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom source method    start      end score strand phase  \\\n",
       "0     1      .  miRNA    17368    17434     .      -     .   \n",
       "1     1      .  miRNA    30365    30501     .      +     .   \n",
       "2     1      .  miRNA   567704   567791     .      -     .   \n",
       "3     1      .  miRNA  1102483  1102576     .      +     .   \n",
       "4     1      .  miRNA  1103242  1103330     .      +     .   \n",
       "\n",
       "                                   group  \n",
       "0  ACC=\"MI0022705\"; ID=\"hsa-mir-6859-1\";  \n",
       "1  ACC=\"MI0006363\"; ID=\"hsa-mir-1302-2\";  \n",
       "2    ACC=\"MI0022558\"; ID=\"hsa-mir-6723\";  \n",
       "3    ACC=\"MI0000342\"; ID=\"hsa-mir-200b\";  \n",
       "4    ACC=\"MI0000737\"; ID=\"hsa-mir-200a\";  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mir_base = pandas.read_table(data_dir / 'hsa.gff2.txt', skiprows=8, names=['chrom', 'source', \n",
    "                                                                           'method', 'start', 'end', \n",
    "                                                                           'score', 'strand',\n",
    "                                                                          'phase', 'group'])\n",
    "# turn zero based:\n",
    "mir_base['start'] -= 1\n",
    "mir_base['end'] -= 2\n",
    "mir_base['chrom'] = mir_base.chrom.str.replace('chr', '')\n",
    "mir_base.to_hdf(results_dir / 'mir_base.hdf', 'df', format='table', mode='w')\n",
    "mir_base.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Ensembl biomart genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000261657</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>66119285</td>\n",
       "      <td>66465398</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SLC25A26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223116</td>\n",
       "      <td>13</td>\n",
       "      <td>23551994</td>\n",
       "      <td>23552136</td>\n",
       "      <td>-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>AL157931.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000233440</td>\n",
       "      <td>13</td>\n",
       "      <td>23708313</td>\n",
       "      <td>23708703</td>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>HMGA1P6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000207157</td>\n",
       "      <td>13</td>\n",
       "      <td>23726725</td>\n",
       "      <td>23726825</td>\n",
       "      <td>-1</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>RNY3P4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000229483</td>\n",
       "      <td>13</td>\n",
       "      <td>23743974</td>\n",
       "      <td>23744736</td>\n",
       "      <td>-1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>LINC00362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID        chrom     start       end  strand       Gene type  \\\n",
       "0  ENSG00000261657  HG991_PATCH  66119285  66465398       1  protein_coding   \n",
       "1  ENSG00000223116           13  23551994  23552136      -1           miRNA   \n",
       "2  ENSG00000233440           13  23708313  23708703       1      pseudogene   \n",
       "3  ENSG00000207157           13  23726725  23726825      -1        misc_RNA   \n",
       "4  ENSG00000229483           13  23743974  23744736      -1         lincRNA   \n",
       "\n",
       "         name  \n",
       "0    SLC25A26  \n",
       "1  AL157931.1  \n",
       "2     HMGA1P6  \n",
       "3      RNY3P4  \n",
       "4   LINC00362  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart_genes = pandas.read_table(data_dir / 'biomart_genes_hg19.tsv')\n",
    "biomart_genes.rename(columns={'Gene name': 'name', \n",
    "                   'Chromosome/scaffold name': 'chrom',\n",
    "                   'Gene start (bp)': 'start', \n",
    "                   'Gene end (bp)': 'end', \n",
    "                   'Strand': 'strand'}, inplace=True)\n",
    "biomart_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomart_genes.to_hdf(results_dir / 'biomart_genes.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensembl biomart genes with relevant GO ontologies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['stable_id', 'start', 'end',  'strand',  'name', 'go_acc', 'go_term', 'chrom']\n",
    "    \n",
    "biomart_chrX_hg19_go_meiosis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_meiosis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_meiosis.to_hdf(results_dir / 'biomart_chrX_hg19_go_meiosis.hdf', 'df', format='table', mode='w')\n",
    "                                          \n",
    "biomart_chrX_hg19_go_spermatogenesis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_spermatogenesis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_spermatogenesis.to_hdf(results_dir / 'biomart_chrX_hg19_go_spermatogenesis.hdf', 'df', format='table', mode='w')\n",
    "    \n",
    "biomart_chrX_hg19_go_apoptosis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_apoptosis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_apoptosis.to_hdf(results_dir / 'biomart_chrX_hg19_go_apoptosis.hdf', 'df', format='table', mode='w')\n",
    "    \n",
    "# biomart_genes_with_go = pandas.read_table(str(data_dir / 'biomart_genes_hg19_genes_with_releveant_go.txt'),\n",
    "#                         header=0, names=['stable_id', 'chrom', 'start', 'end', 'strand', 'name', 'go_acc', 'go_term'])\n",
    "# biomart_genes_with_go.to_hdf(results_dir / 'biomart_genes_with_go.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phastcons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "phastcons_windows = pandas.read_table(data_dir / 'phastCons46way.primates.100kb.tsv',\n",
    "                 header=0, dtype={'chrom': str}, names=['chrom', 'start', 'end', 'phastcons', 'var', 'skew'])\n",
    "phastcons_windows.chrom = [x.replace('chr', '') for x in phastcons_windows.chrom]\n",
    "phastcons_windows = phastcons_windows[['chrom', 'start', 'end', 'phastcons']]\n",
    "\n",
    "phastcons_windows.to_hdf(results_dir / 'phastcons_windows.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Akey 2009 sweep meta analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "akey_regions = (pandas.read_table(str(data_dir / 'Akey09_TableS1_hg19.tsv'),\n",
    "                                        header=0,\n",
    "                                        names=['chrom', 'start', 'end', 'nr_scans', 'references']\n",
    "                                       )\n",
    "                      .loc[:, ['chrom', 'start', 'end']]\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "\n",
    "akey_regions.to_hdf(results_dir / 'akey_regions.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load great ape pi and ILS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pi data from Kiwoong's project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_list = [pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_C.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_G.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_O.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_B.txt'))]\n",
    "df = pandas.concat(df_list)[['species', 'chr', 'pos', 'n.total', 'pi']]\n",
    "df.columns = ['species', 'chrom', 'start', 'total', 'pi']\n",
    "df.chrom = [x.replace('chr', '') for x in df.chrom]\n",
    "df.start *= 100000\n",
    "df['end'] = df.start + 100000\n",
    "\n",
    "\n",
    "# def call_low_ape_pi_regions(df, maxpi, min_analyzed):\n",
    "#     df = (df\n",
    "#           .assign(pi = lambda df: df.pi.where(df.total >= min_analyzed)) \n",
    "#           .loc[:, ['chrom', 'start', 'end', 'species', 'pi']]\n",
    "#           .groupby(['chrom', 'start', 'end', 'species'])\n",
    "#           .aggregate(silent_nanmean)\n",
    "#           .reset_index()\n",
    "#           )\n",
    "    \n",
    "#     group_df_list = list()\n",
    "#     for name, group in df.groupby(['chrom', 'species']):\n",
    "#         group_df = (group\n",
    "#                   .reset_index()\n",
    "#                   .assign(run = 1, \n",
    "#                           islow = lambda df: df.pi <= maxpi) # turn masked into nan to break runs at missing data\n",
    "#                   .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "#                   .groupby(['chrom', 'species', 'test'])\n",
    "#                   .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'pi': 'mean'})\n",
    "#                   .reset_index()\n",
    "#                   .loc[lambda df: df.pi <= maxpi, :]\n",
    "#              )\n",
    "#         group_df_list.append(group_df)\n",
    "\n",
    "#     return pandas.concat(group_df_list)\n",
    "    \n",
    "    \n",
    "def call_low_ape_pi_regions(df, min_analyzed):\n",
    "    df = (df\n",
    "          .assign(pi = lambda df: df.pi.where(df.total >= min_analyzed)) \n",
    "          .loc[:, ['chrom', 'start', 'end', 'species', 'pi']]\n",
    "          .groupby(['chrom', 'start', 'end', 'species'])\n",
    "          .aggregate(silent_nanmean)\n",
    "          .reset_index()\n",
    "          )\n",
    "    \n",
    "    group_df_list = list()\n",
    "    for name, group in df.groupby(['chrom', 'species']):\n",
    "        \n",
    "        maxpi = group.pi.mean() * 0.2\n",
    "        \n",
    "        group_df = (group\n",
    "                  .reset_index()\n",
    "                  .assign(run = 1, \n",
    "                          islow = lambda df: df.pi <= maxpi) # turn masked into nan to break runs at missing data\n",
    "                  .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "                  .groupby(['chrom', 'species', 'test'])\n",
    "                  .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'pi': 'mean'})\n",
    "                  .reset_index()\n",
    "                  .loc[lambda df: df.pi <= maxpi, :]\n",
    "             )\n",
    "        group_df_list.append(group_df)\n",
    "\n",
    "    return pandas.concat(group_df_list)\n",
    "\n",
    "df = (df.loc[df.chrom == 'X']\n",
    "          .groupby(['chrom', 'species', 'start', 'end'])\n",
    "          .aggregate(numpy.mean)\n",
    "         )    \n",
    "df = df.reset_index()\n",
    "\n",
    "#max_pi_fraction_in_low_regions_apes = 0.2\n",
    "#min_fraction_analyzed_apes = 30000\n",
    "\n",
    "#low_ape_pi_regions = call_low_ape_pi_regions(df, max_pi_fraction_in_low_regions_apes, min_fraction_analyzed_apes)\n",
    "\n",
    "max_pi_in_low_regions_apes = 0.0002\n",
    "min_fraction_analyzed_apes = 30000\n",
    "\n",
    "#low_ape_pi_regions = call_low_ape_pi_regions(df, max_pi_in_low_regions_apes, min_fraction_analyzed_apes)\n",
    "low_ape_pi_regions = call_low_ape_pi_regions(df, min_fraction_analyzed_apes)\n",
    "\n",
    "low_ape_pi_regions.to_hdf(results_dir / 'low_ape_pi_regions.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>species</th>\n",
       "      <th>test</th>\n",
       "      <th>start</th>\n",
       "      <th>pi</th>\n",
       "      <th>run</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>2</td>\n",
       "      <td>4300000</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2</td>\n",
       "      <td>4500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>4</td>\n",
       "      <td>8800000</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>1</td>\n",
       "      <td>8900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>6</td>\n",
       "      <td>9000000</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>1</td>\n",
       "      <td>9100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>8</td>\n",
       "      <td>10900000</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>1</td>\n",
       "      <td>11000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>10</td>\n",
       "      <td>16000000</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>4</td>\n",
       "      <td>16400000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom species  test     start        pi  run       end\n",
       "1     X      BO     2   4300000  0.000103    2   4500000\n",
       "3     X      BO     4   8800000  0.000038    1   8900000\n",
       "5     X      BO     6   9000000  0.000112    1   9100000\n",
       "7     X      BO     8  10900000  0.000088    1  11000000\n",
       "9     X      BO    10  16000000  0.000064    4  16400000"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_ape_pi_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BO', 'CC', 'EC', 'ELG', 'NC', 'SO', 'WC', 'WLG'], dtype=object)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_ape_pi_regions.species.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute total superset of regions (excl. EWG that was excluded in Nam et al.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.18805883098508822"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from GenomicIntervals import interval_collapse\n",
    "\n",
    "species_included = ['BO', 'CC', 'EC', 'NC', 'SO', 'WC', 'WLG']#, 'ELG']\n",
    "\n",
    "df = interval_collapse(low_ape_pi_regions.loc[lambda df: df.species.isin(species_included)]\n",
    "                       .sort_values(['chrom', 'start', 'end']))\n",
    "\n",
    "df.assign(length=df.end-df.start).length.sum() / 155270560\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proportions of ILS in 1Mb windows from great ape ILS project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "great_ape_ils_analysis_dir = Path('/home/kmt/projects/great_ape_ils_maps/analyses')\n",
    "\n",
    "window_size = 1000000\n",
    "col_names = [\"start\", \"end\", 'top1', 'top2', 'top3', 'top4']\n",
    "\n",
    "human_chimp_chrX_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/results/chr1-22_X.statesBin1000000chrX.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_chimp_chrX_ils = (pandas.read_table(human_chimp_chrX_ils_path, names=col_names)\n",
    "                        .assign(chrom = 'X',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "human_orang_chrX_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_ponAbe2_nomLeu1_rheMac2/results/chr1-22_X.statesBin1000000chrX.import.coalhmm.list_chrHASH-5473120441099014340.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_orang_chrX_ils = (pandas.read_table(human_orang_chrX_ils_path, names=col_names)\n",
    "                        .assign(chrom = 'X',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "\n",
    "human_chimp_chr7_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/results/chr1-22_X.statesBin1000000chr7.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_chimp_chr7_ils = (pandas.read_table(human_chimp_chr7_ils_path, names=col_names)\n",
    "                        .assign(chrom = '7',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "human_orang_chr7_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_ponAbe2_nomLeu1_rheMac2/results/chr1-22_X.statesBin1000000chr7.import.coalhmm.list_chrHASH-5473120441099014340.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_orang_chr7_ils = (pandas.read_table(human_orang_chr7_ils_path, names=col_names)\n",
    "                        .assign(chrom = '7',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "def call_low_ils_regions(df, max_ils, min_analyzed):\n",
    "    df = (df\n",
    "          .assign(prop_ils = lambda df: df.prop_ils.where(df.fraction_analyzed >= min_analyzed)) \n",
    "          .loc[:, ['chrom', 'start', 'end', 'prop_ils']]\n",
    "          .assign(run = 1, \n",
    "                  islow = lambda df: df.prop_ils <= max_ils) # turn masked into nan to break runs at missing data\n",
    "          .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "          .groupby(['chrom', 'test'])\n",
    "                  .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'prop_ils': 'mean'})\n",
    "                  .reset_index()\n",
    "          .loc[lambda df: df.prop_ils <= max_ils, :]\n",
    "         )\n",
    "    return df\n",
    "\n",
    "max_ils_in_low_regions = 0.05\n",
    "minimal_fraction_analyzed_ils = 0.3\n",
    "\n",
    "human_chimp_low_ils_regions_chrX = call_low_ils_regions(human_chimp_chrX_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_chimp_low_ils_regions_chrX.to_hdf(results_dir / 'human_chimp_low_ils_regions_chrX.hdf', 'df', mode='w')\n",
    "\n",
    "human_orang_low_ils_regions_chrX = call_low_ils_regions(human_orang_chrX_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_orang_low_ils_regions_chrX.to_hdf(results_dir / 'human_orang_low_ils_regions_chrX.hdf', 'df', mode='w')\n",
    "\n",
    "\n",
    "human_chimp_low_ils_regions_chr7 = call_low_ils_regions(human_chimp_chr7_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_chimp_low_ils_regions_chr7.to_hdf(results_dir / 'human_chimp_low_ils_regions_chr7.hdf', 'df', mode='w')\n",
    "\n",
    "human_orang_low_ils_regions_chr7 = call_low_ils_regions(human_orang_chr7_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_orang_low_ils_regions_chr7.to_hdf(results_dir / 'human_orang_low_ils_regions_chr7.hdf', 'df', mode='w')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>53100000.0</td>\n",
       "      <td>53200000.0</td>\n",
       "      <td>0.000492</td>\n",
       "      <td>0.000694</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>0.000520</td>\n",
       "      <td>0.000178</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.48738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>53200000.0</td>\n",
       "      <td>53300000.0</td>\n",
       "      <td>0.000915</td>\n",
       "      <td>0.001316</td>\n",
       "      <td>0.000514</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>0.001158</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.22943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>53300000.0</td>\n",
       "      <td>53400000.0</td>\n",
       "      <td>0.000843</td>\n",
       "      <td>0.001150</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000450</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.000393</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.35590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>53400000.0</td>\n",
       "      <td>53500000.0</td>\n",
       "      <td>0.000861</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.39497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>53500000.0</td>\n",
       "      <td>53600000.0</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.001448</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.000997</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>0.000595</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.36300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>53600000.0</td>\n",
       "      <td>53700000.0</td>\n",
       "      <td>0.000943</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.000661</td>\n",
       "      <td>0.000322</td>\n",
       "      <td>0.000490</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.000622</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.46651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>53700000.0</td>\n",
       "      <td>53800000.0</td>\n",
       "      <td>0.000976</td>\n",
       "      <td>0.001320</td>\n",
       "      <td>0.000633</td>\n",
       "      <td>0.000427</td>\n",
       "      <td>0.000659</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.000810</td>\n",
       "      <td>0.000288</td>\n",
       "      <td>0.32775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>53800000.0</td>\n",
       "      <td>53900000.0</td>\n",
       "      <td>0.000892</td>\n",
       "      <td>0.001216</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.000491</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000862</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.33648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>53900000.0</td>\n",
       "      <td>54000000.0</td>\n",
       "      <td>0.000899</td>\n",
       "      <td>0.001144</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.000288</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.000611</td>\n",
       "      <td>0.000813</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.58934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>54000000.0</td>\n",
       "      <td>54100000.0</td>\n",
       "      <td>0.001036</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>0.000762</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000777</td>\n",
       "      <td>0.001015</td>\n",
       "      <td>0.000539</td>\n",
       "      <td>0.54072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>54100000.0</td>\n",
       "      <td>54200000.0</td>\n",
       "      <td>0.001021</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000672</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000810</td>\n",
       "      <td>0.001122</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.33314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>54200000.0</td>\n",
       "      <td>54300000.0</td>\n",
       "      <td>0.001256</td>\n",
       "      <td>0.001672</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>0.001345</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000486</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.28667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>54300000.0</td>\n",
       "      <td>54400000.0</td>\n",
       "      <td>0.001133</td>\n",
       "      <td>0.001471</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.38844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>54400000.0</td>\n",
       "      <td>54500000.0</td>\n",
       "      <td>0.000888</td>\n",
       "      <td>0.001211</td>\n",
       "      <td>0.000565</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.33778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>54500000.0</td>\n",
       "      <td>54600000.0</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.001045</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.27734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>54600000.0</td>\n",
       "      <td>54700000.0</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.000486</td>\n",
       "      <td>0.000732</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.32928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>54700000.0</td>\n",
       "      <td>54800000.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.001120</td>\n",
       "      <td>0.000557</td>\n",
       "      <td>0.000599</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.000396</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.41748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>54800000.0</td>\n",
       "      <td>54900000.0</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.001036</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000232</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.47382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>54900000.0</td>\n",
       "      <td>55000000.0</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.000476</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000197</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.50642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>55000000.0</td>\n",
       "      <td>55100000.0</td>\n",
       "      <td>0.000669</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000429</td>\n",
       "      <td>0.000518</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.46367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551</th>\n",
       "      <td>55100000.0</td>\n",
       "      <td>55200000.0</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.000947</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>0.000239</td>\n",
       "      <td>0.000228</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.39480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>55200000.0</td>\n",
       "      <td>55300000.0</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>0.001031</td>\n",
       "      <td>0.000442</td>\n",
       "      <td>0.000589</td>\n",
       "      <td>0.000854</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.000292</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.33937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>55300000.0</td>\n",
       "      <td>55400000.0</td>\n",
       "      <td>0.000646</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.000746</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.35616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>55400000.0</td>\n",
       "      <td>55500000.0</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.000992</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000468</td>\n",
       "      <td>0.000792</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.19239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>55500000.0</td>\n",
       "      <td>55600000.0</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>0.001139</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000374</td>\n",
       "      <td>0.000740</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.000637</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.13379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>55600000.0</td>\n",
       "      <td>55700000.0</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.000694</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.001009</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.26309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>55700000.0</td>\n",
       "      <td>55800000.0</td>\n",
       "      <td>0.000773</td>\n",
       "      <td>0.001050</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000374</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.40092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>55800000.0</td>\n",
       "      <td>55900000.0</td>\n",
       "      <td>0.000775</td>\n",
       "      <td>0.001099</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.000895</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.29662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>55900000.0</td>\n",
       "      <td>56000000.0</td>\n",
       "      <td>0.000683</td>\n",
       "      <td>0.001087</td>\n",
       "      <td>0.000280</td>\n",
       "      <td>0.000626</td>\n",
       "      <td>0.001015</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.17558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>56000000.0</td>\n",
       "      <td>56100000.0</td>\n",
       "      <td>0.000280</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000280</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.21411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>56100000.0</td>\n",
       "      <td>56200000.0</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.000889</td>\n",
       "      <td>0.000420</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.47343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>56200000.0</td>\n",
       "      <td>56300000.0</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.001033</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.000643</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.49766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>56300000.0</td>\n",
       "      <td>56400000.0</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.000288</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.41659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>56400000.0</td>\n",
       "      <td>56500000.0</td>\n",
       "      <td>0.000323</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>0.000452</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.27854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>56500000.0</td>\n",
       "      <td>56600000.0</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.000839</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.000540</td>\n",
       "      <td>0.000777</td>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.38861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>56600000.0</td>\n",
       "      <td>56700000.0</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.31328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>56700000.0</td>\n",
       "      <td>56800000.0</td>\n",
       "      <td>0.000448</td>\n",
       "      <td>0.000692</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000448</td>\n",
       "      <td>0.000692</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.31236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>56800000.0</td>\n",
       "      <td>56900000.0</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.47140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0           1         2         3         4         5         6   \\\n",
       "531  53100000.0  53200000.0  0.000492  0.000694  0.000291  0.000349  0.000520   \n",
       "532  53200000.0  53300000.0  0.000915  0.001316  0.000514  0.000785  0.001158   \n",
       "533  53300000.0  53400000.0  0.000843  0.001150  0.000536  0.000450  0.000677   \n",
       "534  53400000.0  53500000.0  0.000861  0.001155  0.000567  0.000430  0.000641   \n",
       "535  53500000.0  53600000.0  0.001102  0.001448  0.000756  0.000716  0.000997   \n",
       "536  53600000.0  53700000.0  0.000943  0.001225  0.000661  0.000322  0.000490   \n",
       "537  53700000.0  53800000.0  0.000976  0.001320  0.000633  0.000427  0.000659   \n",
       "538  53800000.0  53900000.0  0.000892  0.001216  0.000567  0.000297  0.000491   \n",
       "539  53900000.0  54000000.0  0.000899  0.001144  0.000655  0.000288  0.000430   \n",
       "540  54000000.0  54100000.0  0.001036  0.001309  0.000762  0.000259  0.000400   \n",
       "541  54100000.0  54200000.0  0.001021  0.001369  0.000672  0.000210  0.000378   \n",
       "542  54200000.0  54300000.0  0.001256  0.001672  0.000840  0.000977  0.001345   \n",
       "543  54300000.0  54400000.0  0.001133  0.001471  0.000794  0.000566  0.000809   \n",
       "544  54400000.0  54500000.0  0.000888  0.001211  0.000565  0.000444  0.000677   \n",
       "545  54500000.0  54600000.0  0.000721  0.001045  0.000397  0.000505  0.000779   \n",
       "546  54600000.0  54700000.0  0.000698  0.000990  0.000407  0.000486  0.000732   \n",
       "547  54700000.0  54800000.0  0.000838  0.001120  0.000557  0.000599  0.000838   \n",
       "548  54800000.0  54900000.0  0.000781  0.001036  0.000526  0.000549  0.000764   \n",
       "549  54900000.0  55000000.0  0.000513  0.000715  0.000312  0.000316  0.000476   \n",
       "550  55000000.0  55100000.0  0.000669  0.000908  0.000429  0.000518  0.000729   \n",
       "551  55100000.0  55200000.0  0.000684  0.000947  0.000421  0.000456  0.000673   \n",
       "552  55200000.0  55300000.0  0.000737  0.001031  0.000442  0.000589  0.000854   \n",
       "553  55300000.0  55400000.0  0.000646  0.000916  0.000376  0.000505  0.000746   \n",
       "554  55400000.0  55500000.0  0.000624  0.000992  0.000255  0.000468  0.000792   \n",
       "555  55500000.0  55600000.0  0.000673  0.001139  0.000207  0.000374  0.000740   \n",
       "556  55600000.0  55700000.0  0.001102  0.001511  0.000694  0.000684  0.001009   \n",
       "557  55700000.0  55800000.0  0.000773  0.001050  0.000497  0.000624  0.000873   \n",
       "558  55800000.0  55900000.0  0.000775  0.001099  0.000451  0.000607  0.000895   \n",
       "559  55900000.0  56000000.0  0.000683  0.001087  0.000280  0.000626  0.001015   \n",
       "560  56000000.0  56100000.0  0.000280  0.000526  0.000035  0.000280  0.000526   \n",
       "561  56100000.0  56200000.0  0.000655  0.000889  0.000420  0.000549  0.000764   \n",
       "562  56200000.0  56300000.0  0.000784  0.001033  0.000534  0.000643  0.000869   \n",
       "563  56300000.0  56400000.0  0.000336  0.000519  0.000153  0.000288  0.000458   \n",
       "564  56400000.0  56500000.0  0.000323  0.000547  0.000099  0.000251  0.000452   \n",
       "565  56500000.0  56600000.0  0.000592  0.000839  0.000345  0.000540  0.000777   \n",
       "566  56600000.0  56700000.0  0.000192  0.000359  0.000024  0.000192  0.000359   \n",
       "567  56700000.0  56800000.0  0.000448  0.000692  0.000205  0.000448  0.000692   \n",
       "568  56800000.0  56900000.0  0.000085  0.000181  0.000000  0.000085  0.000181   \n",
       "\n",
       "           7         8         9         10       11  \n",
       "531  0.000178  0.000144  0.000259  0.000029  0.48738  \n",
       "532  0.000412  0.000131  0.000312  0.000000  0.22943  \n",
       "533  0.000222  0.000393  0.000607  0.000180  0.35590  \n",
       "534  0.000220  0.000430  0.000641  0.000220  0.39497  \n",
       "535  0.000435  0.000386  0.000595  0.000176  0.36300  \n",
       "536  0.000153  0.000622  0.000852  0.000391  0.46651  \n",
       "537  0.000195  0.000549  0.000810  0.000288  0.32775  \n",
       "538  0.000103  0.000594  0.000862  0.000327  0.33648  \n",
       "539  0.000147  0.000611  0.000813  0.000408  0.58934  \n",
       "540  0.000118  0.000777  0.001015  0.000539  0.54072  \n",
       "541  0.000042  0.000810  0.001122  0.000499  0.33314  \n",
       "542  0.000608  0.000279  0.000486  0.000072  0.28667  \n",
       "543  0.000324  0.000566  0.000809  0.000324  0.38844  \n",
       "544  0.000211  0.000444  0.000677  0.000211  0.33778  \n",
       "545  0.000230  0.000216  0.000406  0.000027  0.27734  \n",
       "546  0.000240  0.000213  0.000383  0.000042  0.32928  \n",
       "547  0.000359  0.000240  0.000396  0.000083  0.41748  \n",
       "548  0.000334  0.000232  0.000376  0.000088  0.47382  \n",
       "549  0.000156  0.000197  0.000326  0.000068  0.50642  \n",
       "550  0.000306  0.000151  0.000272  0.000030  0.46367  \n",
       "551  0.000239  0.000228  0.000386  0.000070  0.39480  \n",
       "552  0.000324  0.000147  0.000292  0.000003  0.33937  \n",
       "553  0.000265  0.000140  0.000278  0.000003  0.35616  \n",
       "554  0.000144  0.000156  0.000372  0.000000  0.19239  \n",
       "555  0.000007  0.000299  0.000637  0.000000  0.13379  \n",
       "556  0.000359  0.000418  0.000677  0.000159  0.26309  \n",
       "557  0.000374  0.000150  0.000281  0.000018  0.40092  \n",
       "558  0.000318  0.000169  0.000334  0.000003  0.29662  \n",
       "559  0.000238  0.000057       NaN       NaN  0.17558  \n",
       "560  0.000035  0.000000       NaN       NaN  0.21411  \n",
       "561  0.000334  0.000106  0.000209  0.000002  0.47343  \n",
       "562  0.000417  0.000141  0.000253  0.000028  0.49766  \n",
       "563  0.000118  0.000048  0.000142  0.000000  0.41659  \n",
       "564  0.000050  0.000072  0.000213  0.000000  0.27854  \n",
       "565  0.000304  0.000051  0.000152  0.000000  0.38861  \n",
       "566  0.000024  0.000000       NaN       NaN  0.31328  \n",
       "567  0.000205  0.000000       NaN       NaN  0.31236  \n",
       "568  0.000000  0.000000       NaN       NaN  0.47140  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/home/kmt/projects/great_ape_ils_maps/analyses/multiz_hg19_panTro3_gorGor3_ponAbe2/results/chr1-22_X.bm100000.minp050maxsinfchr17.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl'\n",
    "\n",
    "human_orang_chr9_100kb = (pandas.read_table(f, header=None)#, names=col_names)\n",
    "#                         .assign(chrom = '9',\n",
    "#                                 prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "#                                 fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "#human_orang_chr9_ils\n",
    "human_orang_chr9_100kb.loc[lambda df: (df[0] > 53e6) & (df[1] < 57e6)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>test</th>\n",
       "      <th>run</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>prop_ils</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>26000000</td>\n",
       "      <td>29000000</td>\n",
       "      <td>0.041745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>40000000</td>\n",
       "      <td>43000000</td>\n",
       "      <td>0.042974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>X</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>68000000</td>\n",
       "      <td>70000000</td>\n",
       "      <td>0.033776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>X</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>71000000</td>\n",
       "      <td>72000000</td>\n",
       "      <td>0.048062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>98000000</td>\n",
       "      <td>101000000</td>\n",
       "      <td>0.023426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>X</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>106000000</td>\n",
       "      <td>108000000</td>\n",
       "      <td>0.014441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>124000000</td>\n",
       "      <td>125000000</td>\n",
       "      <td>0.036677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>X</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>129000000</td>\n",
       "      <td>134000000</td>\n",
       "      <td>0.038673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>X</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>154000000</td>\n",
       "      <td>155000000</td>\n",
       "      <td>0.035043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom  test  run      start        end  prop_ils\n",
       "1      X     2    3   26000000   29000000  0.041745\n",
       "3      X     4    3   40000000   43000000  0.042974\n",
       "5      X     6    2   68000000   70000000  0.033776\n",
       "7      X     8    1   71000000   72000000  0.048062\n",
       "9      X    10    3   98000000  101000000  0.023426\n",
       "11     X    12    2  106000000  108000000  0.014441\n",
       "13     X    14    1  124000000  125000000  0.036677\n",
       "15     X    16    5  129000000  134000000  0.038673\n",
       "17     X    18    1  154000000  155000000  0.035043"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = call_low_ils_regions(human_orang_chrX_ils, \n",
    "                           max_ils_in_low_regions, \n",
    "                           minimal_fraction_analyzed_ils)\n",
    "\n",
    "df#.loc[(df.start > 50e6) & (df.end < 60e6)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.13332299281536769"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_chimp_chrX_ils.loc[human_chimp_chrX_ils.fraction_analyzed >= minimal_fraction_analyzed_ils].prop_ils.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.11592667663464343"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = (human_chimp_chrX_ils\n",
    "      .loc[human_chimp_chrX_ils.fraction_analyzed >= minimal_fraction_analyzed_ils]\n",
    "      .loc[human_chimp_chrX_ils.prop_ils < 0.02]\n",
    "     )\n",
    "len(df) / 155.270560"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Trine Line's circRNA results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trine_line_genes = pandas.read_csv(data_dir / 'trine_line_data/AnnoTable.csv')\n",
    "trine_line_genes.drop(trine_line_genes.columns[[0]], axis=1, inplace=True)\n",
    "trine_line_x_genes = trine_line_genes.loc[trine_line_genes.chr == 'chrX']\n",
    "trine_line_x_genes.to_hdf(results_dir / 'trine_line_x_genes.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chalmel genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read gene table from Chalmel paper SOM (I exported csv from the excel table):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Description</th>\n",
       "      <th>biological_process terms</th>\n",
       "      <th>molecular_function terms</th>\n",
       "      <th>cellular_component terms</th>\n",
       "      <th>Human_Phenotype</th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Expression in Testis</th>\n",
       "      <th>TF</th>\n",
       "      <th>...</th>\n",
       "      <th>tonsil-non_germinal_center_cells</th>\n",
       "      <th>tonsil-squamous_epithelial_cells</th>\n",
       "      <th>urinary_bladder-urothelial_cells</th>\n",
       "      <th>uterus_post_menopause-cells_in_endometrial_stroma</th>\n",
       "      <th>uterus_post_menopause-glandular_cells</th>\n",
       "      <th>uterus_pre_menopause-cells_in_endometrial_stroma</th>\n",
       "      <th>uterus_pre_menopause-glandular_cells</th>\n",
       "      <th>vagina-squamous_epithelial_cells</th>\n",
       "      <th>vulva_anal_skin-epidermal_cells</th>\n",
       "      <th>Unnamed: 180</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>210262_at</td>\n",
       "      <td>CRISP2</td>\n",
       "      <td>cysteine-rich secretory protein 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>extracellular space (0005615)</td>\n",
       "      <td>-</td>\n",
       "      <td>12</td>\n",
       "      <td>PEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>206112_at</td>\n",
       "      <td>ANKRD7</td>\n",
       "      <td>ankyrin repeat domain 7</td>\n",
       "      <td>male gonad development (0008584)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>11</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1558262_at</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>207116_s_at</td>\n",
       "      <td>GAPDHS</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase, sper...</td>\n",
       "      <td>gluconeogenesis (0006094);glycolysis (0006096)...</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase (NAD+...</td>\n",
       "      <td>cytosol (0005829);microtubule-based flagellum ...</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>231612_at</td>\n",
       "      <td>C4orf35</td>\n",
       "      <td>chromosome 4 open reading frame 35</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 181 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID GeneName                                        Description  \\\n",
       "0    210262_at   CRISP2                  cysteine-rich secretory protein 2   \n",
       "1    206112_at   ANKRD7                            ankyrin repeat domain 7   \n",
       "2   1558262_at        -                                                  -   \n",
       "3  207116_s_at   GAPDHS  glyceraldehyde-3-phosphate dehydrogenase, sper...   \n",
       "4    231612_at  C4orf35                 chromosome 4 open reading frame 35   \n",
       "\n",
       "                            biological_process terms  \\\n",
       "0                                                  -   \n",
       "1                   male gonad development (0008584)   \n",
       "2                                                  -   \n",
       "3  gluconeogenesis (0006094);glycolysis (0006096)...   \n",
       "4                                                  -   \n",
       "\n",
       "                            molecular_function terms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2                                                  -   \n",
       "3  glyceraldehyde-3-phosphate dehydrogenase (NAD+...   \n",
       "4                                                  -   \n",
       "\n",
       "                            cellular_component terms Human_Phenotype Pattern  \\\n",
       "0                      extracellular space (0005615)               -      12   \n",
       "1                                                  -               -      11   \n",
       "2                                                  -               -      13   \n",
       "3  cytosol (0005829);microtubule-based flagellum ...               -      13   \n",
       "4                                                  -               -      13   \n",
       "\n",
       "  Expression in Testis TF      ...      tonsil-non_germinal_center_cells  \\\n",
       "0                PEHET  -      ...                                     0   \n",
       "1                SEHET  -      ...                                     -   \n",
       "2                SEHET  -      ...                                     -   \n",
       "3                SEHET  -      ...                                     -   \n",
       "4                SEHET  -      ...                                     -   \n",
       "\n",
       "  tonsil-squamous_epithelial_cells urinary_bladder-urothelial_cells  \\\n",
       "0                                0                                0   \n",
       "1                                -                                -   \n",
       "2                                -                                -   \n",
       "3                                -                                -   \n",
       "4                                -                                -   \n",
       "\n",
       "  uterus_post_menopause-cells_in_endometrial_stroma  \\\n",
       "0                                                 0   \n",
       "1                                                 -   \n",
       "2                                                 -   \n",
       "3                                                 -   \n",
       "4                                                 -   \n",
       "\n",
       "   uterus_post_menopause-glandular_cells  \\\n",
       "0                                      0   \n",
       "1                                      -   \n",
       "2                                      -   \n",
       "3                                      -   \n",
       "4                                      -   \n",
       "\n",
       "   uterus_pre_menopause-cells_in_endometrial_stroma  \\\n",
       "0                                                 0   \n",
       "1                                                 -   \n",
       "2                                                 -   \n",
       "3                                                 -   \n",
       "4                                                 -   \n",
       "\n",
       "   uterus_pre_menopause-glandular_cells  vagina-squamous_epithelial_cells  \\\n",
       "0                                     0                                 0   \n",
       "1                                     -                                 -   \n",
       "2                                     -                                 -   \n",
       "3                                     -                                 -   \n",
       "4                                     -                                 -   \n",
       "\n",
       "   vulva_anal_skin-epidermal_cells  Unnamed: 180  \n",
       "0                                0           NaN  \n",
       "1                                -           NaN  \n",
       "2                                -           NaN  \n",
       "3                                -           NaN  \n",
       "4                                -           NaN  \n",
       "\n",
       "[5 rows x 181 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chalmel_genes = pandas.read_csv(data_dir / 'chalmel_genes.csv', sep=';')\n",
    "chalmel_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Description</th>\n",
       "      <th>biological_process terms</th>\n",
       "      <th>molecular_function terms</th>\n",
       "      <th>cellular_component terms</th>\n",
       "      <th>Human_Phenotype</th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Expression in Testis</th>\n",
       "      <th>TF</th>\n",
       "      <th>...</th>\n",
       "      <th>vagina-squamous_epithelial_cells</th>\n",
       "      <th>vulva_anal_skin-epidermal_cells</th>\n",
       "      <th>Unnamed: 180</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>210262_at</td>\n",
       "      <td>CRISP2</td>\n",
       "      <td>cysteine-rich secretory protein 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>extracellular space (0005615)</td>\n",
       "      <td>-</td>\n",
       "      <td>12</td>\n",
       "      <td>PEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000124490</td>\n",
       "      <td>6</td>\n",
       "      <td>49660073</td>\n",
       "      <td>49681274</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CRISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>206112_at</td>\n",
       "      <td>ANKRD7</td>\n",
       "      <td>ankyrin repeat domain 7</td>\n",
       "      <td>male gonad development (0008584)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>11</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000106013</td>\n",
       "      <td>7</td>\n",
       "      <td>117854727</td>\n",
       "      <td>117882785</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ANKRD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>207116_s_at</td>\n",
       "      <td>GAPDHS</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase, sper...</td>\n",
       "      <td>gluconeogenesis (0006094);glycolysis (0006096)...</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase (NAD+...</td>\n",
       "      <td>cytosol (0005829);microtubule-based flagellum ...</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000105679</td>\n",
       "      <td>19</td>\n",
       "      <td>36024314</td>\n",
       "      <td>36036218</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GAPDHS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>243208_x_at</td>\n",
       "      <td>ACTL9</td>\n",
       "      <td>actin-like 9</td>\n",
       "      <td>-</td>\n",
       "      <td>protein binding (0005515)</td>\n",
       "      <td>cytoplasm (0005737);cytoskeleton (0005856)</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000263000</td>\n",
       "      <td>HG729_PATCH</td>\n",
       "      <td>8762287</td>\n",
       "      <td>8763706</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACTL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>243208_x_at</td>\n",
       "      <td>ACTL9</td>\n",
       "      <td>actin-like 9</td>\n",
       "      <td>-</td>\n",
       "      <td>protein binding (0005515)</td>\n",
       "      <td>cytoplasm (0005737);cytoskeleton (0005856)</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000181786</td>\n",
       "      <td>19</td>\n",
       "      <td>8807751</td>\n",
       "      <td>8809172</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACTL9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID GeneName                                        Description  \\\n",
       "0    210262_at   CRISP2                  cysteine-rich secretory protein 2   \n",
       "1    206112_at   ANKRD7                            ankyrin repeat domain 7   \n",
       "2  207116_s_at   GAPDHS  glyceraldehyde-3-phosphate dehydrogenase, sper...   \n",
       "3  243208_x_at    ACTL9                                       actin-like 9   \n",
       "4  243208_x_at    ACTL9                                       actin-like 9   \n",
       "\n",
       "                            biological_process terms  \\\n",
       "0                                                  -   \n",
       "1                   male gonad development (0008584)   \n",
       "2  gluconeogenesis (0006094);glycolysis (0006096)...   \n",
       "3                                                  -   \n",
       "4                                                  -   \n",
       "\n",
       "                            molecular_function terms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2  glyceraldehyde-3-phosphate dehydrogenase (NAD+...   \n",
       "3                          protein binding (0005515)   \n",
       "4                          protein binding (0005515)   \n",
       "\n",
       "                            cellular_component terms Human_Phenotype Pattern  \\\n",
       "0                      extracellular space (0005615)               -      12   \n",
       "1                                                  -               -      11   \n",
       "2  cytosol (0005829);microtubule-based flagellum ...               -      13   \n",
       "3         cytoplasm (0005737);cytoskeleton (0005856)               -      13   \n",
       "4         cytoplasm (0005737);cytoskeleton (0005856)               -      13   \n",
       "\n",
       "  Expression in Testis TF   ...   vagina-squamous_epithelial_cells  \\\n",
       "0                PEHET  -   ...                                  0   \n",
       "1                SEHET  -   ...                                  -   \n",
       "2                SEHET  -   ...                                  -   \n",
       "3                SEHET  -   ...                                  2   \n",
       "4                SEHET  -   ...                                  2   \n",
       "\n",
       "  vulva_anal_skin-epidermal_cells Unnamed: 180   Gene stable ID        chrom  \\\n",
       "0                               0          NaN  ENSG00000124490            6   \n",
       "1                               -          NaN  ENSG00000106013            7   \n",
       "2                               -          NaN  ENSG00000105679           19   \n",
       "3                               2          NaN  ENSG00000263000  HG729_PATCH   \n",
       "4                               2          NaN  ENSG00000181786           19   \n",
       "\n",
       "       start        end  strand       Gene type    name  \n",
       "0   49660073   49681274      -1  protein_coding  CRISP2  \n",
       "1  117854727  117882785       1  protein_coding  ANKRD7  \n",
       "2   36024314   36036218       1  protein_coding  GAPDHS  \n",
       "3    8762287    8763706      -1  protein_coding   ACTL9  \n",
       "4    8807751    8809172      -1  protein_coding   ACTL9  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chalmel_genes = chalmel_genes.merge(biomart_genes, left_on=['GeneName'], right_on=['name'], how='inner')\n",
    "chalmel_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#chalmel_genes.loc[(chalmel_genes.chrom == '3') & (chalmel_genes.start > 45e6) & (chalmel_genes.end < 55e6)].sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chalmel_genes.to_hdf(results_dir / 'chalmel_genes.hdf', 'df', format='table', mode='w')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
