{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Complilation of auxiliary data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, os, sys#, math, pickle, subprocess, resource, random\n",
    "from pathlib import Path\n",
    "import numpy\n",
    "import pandas\n",
    "from pandas import DataFrame, Series\n",
    "\n",
    "numpy.random.seed(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = Path(os.environ['HOME'], 'simons/faststorage/people/kmt')\n",
    "#meta_data_dir = Path(os.environ['HOME'], 'simons/faststorage/data/metadata')\n",
    "#steps_dir = root_dir / 'steps'\n",
    "#argweaver_dir = steps_dir / 'argweaver/output'\n",
    "results_dir = root_dir / 'results'\n",
    "#figures_dir = root_dir / 'figures'\n",
    "data_dir = root_dir / 'data'\n",
    "#pi_dir = steps_dir / 'pi_stores'\n",
    "#dist_dir = steps_dir / 'dist_stores'\n",
    "#male_x_haploid_dir = steps_dir / 'male_x_haploids'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local code in the scripts dir on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    }
   ],
   "source": [
    "scripts_dir = root_dir / 'scripts'\n",
    "if str(scripts_dir) not in sys.path:\n",
    "    sys.path.append(str(scripts_dir))\n",
    "\n",
    "import simons_meta_data\n",
    "import hg19_chrom_sizes\n",
    "\n",
    "from ChromosomeWindows import window\n",
    "import genominterv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convenience functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def silent_nanmean(x):\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        return numpy.nanmean(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The savefig.jpeg_quality rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The keymap.all_axes rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_path rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "In /home/kmt/anaconda3/envs/simons/lib/python3.6/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle: \n",
      "The animation.avconv_args rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# Make inline plots vector graphics instead of raster graphics\n",
    "from IPython.display import set_matplotlib_formats\n",
    "set_matplotlib_formats('pdf', 'svg')\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.patches import Rectangle\n",
    "# from mpl_toolkits.basemap import Basemap\n",
    "#matplotlib.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "\n",
    "import mpld3\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_style(\"whitegrid\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "# lowess for plotting\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "from scipy.interpolate import UnivariateSpline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# silence deprecation warnings (lots from seaborn)\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning) \n",
    "warnings.filterwarnings(\"ignore\", category=numpy.VisibleDeprecationWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load genome annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ampliconic regions and Shriram's regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ampliconic_regions = (pandas.read_table(str(data_dir / 'coordinates_hg18_hg19_hg38_Amplicons_Gap.txt'))\n",
    "                      .assign(start=lambda df: df.hg19_start, \n",
    "                              end=lambda df: df.hg19_end,\n",
    "                              chrom='X')\n",
    "                      .loc[:, ['chrom', 'start', 'end']]\n",
    "                     )\n",
    "ampliconic_regions.head()\n",
    "\n",
    "# ampliconic_regions = (pandas.read_table(str(data_dir / 'AmpliconicRegions_hg19.csv'),\n",
    "#                                        header=0,\n",
    "#                                         names=['chrom', 'start', 'end', 'region', 'midpoint',\n",
    "#                                                         'size', 'alignment', 'pnilsr']\n",
    "#                                        )\n",
    "#                       .loc[:, ['chrom', 'start', 'end']]\n",
    "#                       .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "#                      )\n",
    "\n",
    "ampliconic_regions.to_hdf(results_dir / 'ampliconic_regions.hdf', 'df', mode='w')\n",
    "\n",
    "no_neanderthal_asn = (pandas.read_table(str(data_dir / 'Neanderthal_ASN.csv'), sep=' ', \n",
    "                                       header=0, names=['chrom', 'start', 'end'])\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "no_neanderthal_asn.to_hdf(results_dir / 'no_neanderthal_asn.hdf', 'df', mode='w')\n",
    "\n",
    "\n",
    "no_neanderthal_eur = (pandas.read_table(str(data_dir / 'Neanderthal_EUR.csv'), sep=' ',\n",
    "                                       header=0, names=['chrom', 'start', 'end'])\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "no_neanderthal_eur.to_hdf(results_dir / 'no_neanderthal_eur.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RefSeq genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_genes = (pandas.read_table(str(data_dir / 'refseq_genes_hg19_table.txt')))\n",
    "refseq_genes.chrom = [x.replace('chr', '') for x in refseq_genes.chrom]\n",
    "refseq_genes.rename(columns={\"txStart\": \"start\", \"txEnd\": 'end'}, inplace=True)\n",
    "refseq_genes.sort_values(by=['chrom', 'start', 'end'], inplace=True)\n",
    "\n",
    "refseq_genes.to_hdf(results_dir / 'refseq_genes.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Density of transcribed Refseq gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>100000</td>\n",
       "      <td>0.08090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100000</td>\n",
       "      <td>200000</td>\n",
       "      <td>0.05794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200000</td>\n",
       "      <td>300000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300000</td>\n",
       "      <td>400000</td>\n",
       "      <td>0.05629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>400000</td>\n",
       "      <td>500000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end  gene_density\n",
       "0     1       0  100000       0.08090\n",
       "1     1  100000  200000       0.05794\n",
       "2     1  200000  300000       0.00000\n",
       "3     1  300000  400000       0.05629\n",
       "4     1  400000  500000       0.00000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "window_size = 100000\n",
    "@window(size=window_size, fill='hg19')\n",
    "def gene_density(df):\n",
    "    if not len(df):\n",
    "        return 0\n",
    "    df = genominterv.interval_collapse(df)\n",
    "    return sum(df.end - df.start) / window_size\n",
    "\n",
    "gene_density_df = refseq_genes.groupby('chrom').apply(gene_density).reset_index(level=['chrom'])\n",
    "gene_density_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Density of refseq exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>100000</td>\n",
       "      <td>0.05627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100000</td>\n",
       "      <td>200000</td>\n",
       "      <td>0.05474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>200000</td>\n",
       "      <td>300000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>300000</td>\n",
       "      <td>400000</td>\n",
       "      <td>0.05309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>400000</td>\n",
       "      <td>500000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end  exon_density\n",
       "0     1       0  100000       0.05627\n",
       "1     1  100000  200000       0.05474\n",
       "2     1  200000  300000       0.00000\n",
       "3     1  300000  400000       0.05309\n",
       "4     1  400000  500000       0.00000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_exons(df):\n",
    "    exon_starts = list()\n",
    "    exon_ends = list()\n",
    "    for row in df.itertuples():\n",
    "        chrom = row.chrom\n",
    "        exon_starts.extend(list(map(int, row.exonStarts.split(',')[:-1])))\n",
    "        exon_ends.extend(list(map(int, row.exonEnds.split(',')[:-1])))\n",
    "    all_exons = DataFrame({'chrom': chrom, 'start': exon_starts, 'end': exon_ends})\n",
    "    all_exons.sort_values(['start', 'end'], inplace=True)\n",
    "    collapsed_exons = genominterv.interval_collapse(all_exons)\n",
    "    return collapsed_exons\n",
    "\n",
    "# data frame with exon start end coordinates\n",
    "refseq_exons = refseq_genes.groupby('chrom').apply(get_exons).reset_index(drop=True)\n",
    "\n",
    "window_size = 100000\n",
    "@window(size=window_size)\n",
    "def exon_density(df):\n",
    "    if not len(df):\n",
    "        return 0\n",
    "    return sum(df.end - df.start) / window_size\n",
    "\n",
    "# windowed exon density\n",
    "exon_density_df = refseq_exons.groupby('chrom').apply(exon_density).reset_index(level=['chrom'])\n",
    "exon_density_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge both into one data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_density = pandas.merge(exon_density_df, gene_density_df, on=['chrom', 'start', 'end'])\n",
    "\n",
    "refseq_density.to_hdf(results_dir / 'refseq_density.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mirBase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GFF2 in hg19 ftp://mirbase.org/pub/mirbase/20/genomes/hsa.gff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>method</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17368</td>\n",
       "      <td>17434</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0022705\"; ID=\"hsa-mir-6859-1\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30365</td>\n",
       "      <td>30501</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0006363\"; ID=\"hsa-mir-1302-2\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>567704</td>\n",
       "      <td>567791</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0022558\"; ID=\"hsa-mir-6723\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>1102483</td>\n",
       "      <td>1102576</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0000342\"; ID=\"hsa-mir-200b\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>1103242</td>\n",
       "      <td>1103330</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ACC=\"MI0000737\"; ID=\"hsa-mir-200a\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom source method    start      end score strand phase  \\\n",
       "0     1      .  miRNA    17368    17434     .      -     .   \n",
       "1     1      .  miRNA    30365    30501     .      +     .   \n",
       "2     1      .  miRNA   567704   567791     .      -     .   \n",
       "3     1      .  miRNA  1102483  1102576     .      +     .   \n",
       "4     1      .  miRNA  1103242  1103330     .      +     .   \n",
       "\n",
       "                                   group  \n",
       "0  ACC=\"MI0022705\"; ID=\"hsa-mir-6859-1\";  \n",
       "1  ACC=\"MI0006363\"; ID=\"hsa-mir-1302-2\";  \n",
       "2    ACC=\"MI0022558\"; ID=\"hsa-mir-6723\";  \n",
       "3    ACC=\"MI0000342\"; ID=\"hsa-mir-200b\";  \n",
       "4    ACC=\"MI0000737\"; ID=\"hsa-mir-200a\";  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mir_base = pandas.read_table(data_dir / 'hsa.gff2.txt', skiprows=8, names=['chrom', 'source', \n",
    "                                                                           'method', 'start', 'end', \n",
    "                                                                           'score', 'strand',\n",
    "                                                                          'phase', 'group'])\n",
    "# turn zero based:\n",
    "mir_base['start'] -= 1\n",
    "mir_base['end'] -= 2\n",
    "mir_base['chrom'] = mir_base.chrom.str.replace('chr', '')\n",
    "mir_base.to_hdf(results_dir / 'mir_base.hdf', 'df', format='table', mode='w')\n",
    "mir_base.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Ensembl biomart genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000261657</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>66119285</td>\n",
       "      <td>66465398</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SLC25A26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223116</td>\n",
       "      <td>13</td>\n",
       "      <td>23551994</td>\n",
       "      <td>23552136</td>\n",
       "      <td>-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>AL157931.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000233440</td>\n",
       "      <td>13</td>\n",
       "      <td>23708313</td>\n",
       "      <td>23708703</td>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>HMGA1P6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000207157</td>\n",
       "      <td>13</td>\n",
       "      <td>23726725</td>\n",
       "      <td>23726825</td>\n",
       "      <td>-1</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>RNY3P4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000229483</td>\n",
       "      <td>13</td>\n",
       "      <td>23743974</td>\n",
       "      <td>23744736</td>\n",
       "      <td>-1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>LINC00362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID        chrom     start       end  strand       Gene type  \\\n",
       "0  ENSG00000261657  HG991_PATCH  66119285  66465398       1  protein_coding   \n",
       "1  ENSG00000223116           13  23551994  23552136      -1           miRNA   \n",
       "2  ENSG00000233440           13  23708313  23708703       1      pseudogene   \n",
       "3  ENSG00000207157           13  23726725  23726825      -1        misc_RNA   \n",
       "4  ENSG00000229483           13  23743974  23744736      -1         lincRNA   \n",
       "\n",
       "         name  \n",
       "0    SLC25A26  \n",
       "1  AL157931.1  \n",
       "2     HMGA1P6  \n",
       "3      RNY3P4  \n",
       "4   LINC00362  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart_genes = pandas.read_table(data_dir / 'biomart_genes_hg19.tsv')\n",
    "biomart_genes.rename(columns={'Gene name': 'name', \n",
    "                   'Chromosome/scaffold name': 'chrom',\n",
    "                   'Gene start (bp)': 'start', \n",
    "                   'Gene end (bp)': 'end', \n",
    "                   'Strand': 'strand'}, inplace=True)\n",
    "biomart_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomart_genes.to_hdf(results_dir / 'biomart_genes.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensembl biomart genes with relevant GO ontologies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['stable_id', 'start', 'end',  'strand',  'name', 'go_acc', 'go_term', 'chrom']\n",
    "    \n",
    "biomart_chrX_hg19_go_meiosis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_meiosis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_meiosis.to_hdf(results_dir / 'biomart_chrX_hg19_go_meiosis.hdf', 'df', format='table', mode='w')\n",
    "                                          \n",
    "biomart_chrX_hg19_go_spermatogenesis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_spermatogenesis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_spermatogenesis.to_hdf(results_dir / 'biomart_chrX_hg19_go_spermatogenesis.hdf', 'df', format='table', mode='w')\n",
    "    \n",
    "biomart_chrX_hg19_go_apoptosis = pandas.read_table(str(data_dir / 'biomart_chrX_hg19_go_apoptosis.txt'),\n",
    "                        header=0, names=names)\n",
    "biomart_chrX_hg19_go_apoptosis.to_hdf(results_dir / 'biomart_chrX_hg19_go_apoptosis.hdf', 'df', format='table', mode='w')\n",
    "    \n",
    "# biomart_genes_with_go = pandas.read_table(str(data_dir / 'biomart_genes_hg19_genes_with_releveant_go.txt'),\n",
    "#                         header=0, names=['stable_id', 'chrom', 'start', 'end', 'strand', 'name', 'go_acc', 'go_term'])\n",
    "# biomart_genes_with_go.to_hdf(results_dir / 'biomart_genes_with_go.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phastcons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "phastcons_windows = pandas.read_table(data_dir / 'phastCons46way.primates.100kb.tsv',\n",
    "                 header=0, dtype={'chrom': str}, names=['chrom', 'start', 'end', 'phastcons', 'var', 'skew'])\n",
    "phastcons_windows.chrom = [x.replace('chr', '') for x in phastcons_windows.chrom]\n",
    "phastcons_windows = phastcons_windows[['chrom', 'start', 'end', 'phastcons']]\n",
    "\n",
    "phastcons_windows.to_hdf(results_dir / 'phastcons_windows.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Akey 2009 sweep meta analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "akey_regions = (pandas.read_table(str(data_dir / 'Akey09_TableS1_hg19.tsv'),\n",
    "                                        header=0,\n",
    "                                        names=['chrom', 'start', 'end', 'nr_scans', 'references']\n",
    "                                       )\n",
    "                      .loc[:, ['chrom', 'start', 'end']]\n",
    "                      .assign(chrom = lambda df: [x.replace('chr', '') for x in df.chrom])\n",
    "                     )\n",
    "\n",
    "akey_regions.to_hdf(results_dir / 'akey_regions.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load great ape pi and ILS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pi data from Kiwoong's project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = [pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_C.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_G.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_O.txt')),\n",
    "           pandas.read_table(str(data_dir / 'KiwoongsPiTables/PFW_B.txt'))]\n",
    "df = pandas.concat(df_list)[['species', 'chr', 'pos', 'n.total', 'pi']]\n",
    "df.columns = ['species', 'chrom', 'start', 'total', 'pi']\n",
    "df.chrom = [x.replace('chr', '') for x in df.chrom]\n",
    "df.start *= 100000\n",
    "df['end'] = df.start + 100000\n",
    "\n",
    "\n",
    "# def call_low_ape_pi_regions(df, maxpi, min_analyzed):\n",
    "#     df = (df\n",
    "#           .assign(pi = lambda df: df.pi.where(df.total >= min_analyzed)) \n",
    "#           .loc[:, ['chrom', 'start', 'end', 'species', 'pi']]\n",
    "#           .groupby(['chrom', 'start', 'end', 'species'])\n",
    "#           .aggregate(silent_nanmean)\n",
    "#           .reset_index()\n",
    "#           )\n",
    "    \n",
    "#     group_df_list = list()\n",
    "#     for name, group in df.groupby(['chrom', 'species']):\n",
    "#         group_df = (group\n",
    "#                   .reset_index()\n",
    "#                   .assign(run = 1, \n",
    "#                           islow = lambda df: df.pi <= maxpi) # turn masked into nan to break runs at missing data\n",
    "#                   .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "#                   .groupby(['chrom', 'species', 'test'])\n",
    "#                   .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'pi': 'mean'})\n",
    "#                   .reset_index()\n",
    "#                   .loc[lambda df: df.pi <= maxpi, :]\n",
    "#              )\n",
    "#         group_df_list.append(group_df)\n",
    "\n",
    "#     return pandas.concat(group_df_list)\n",
    "    \n",
    "    \n",
    "def call_low_ape_pi_regions(df, min_analyzed):\n",
    "    df = (df\n",
    "          .assign(pi = lambda df: df.pi.where(df.total >= min_analyzed)) \n",
    "          .loc[:, ['chrom', 'start', 'end', 'species', 'pi']]\n",
    "          .groupby(['chrom', 'start', 'end', 'species'])\n",
    "          .aggregate(silent_nanmean)\n",
    "          .reset_index()\n",
    "          )\n",
    "    \n",
    "    group_df_list = list()\n",
    "    for name, group in df.groupby(['chrom', 'species']):\n",
    "        \n",
    "        maxpi = group.pi.mean() * 0.2\n",
    "        \n",
    "        group_df = (group\n",
    "                  .reset_index()\n",
    "                  .assign(run = 1, \n",
    "                          islow = lambda df: df.pi <= maxpi) # turn masked into nan to break runs at missing data\n",
    "                  .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "                  .groupby(['chrom', 'species', 'test'])\n",
    "                  .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'pi': 'mean'})\n",
    "                  .reset_index()\n",
    "                  .loc[lambda df: df.pi <= maxpi, :]\n",
    "             )\n",
    "        group_df_list.append(group_df)\n",
    "\n",
    "    return pandas.concat(group_df_list)\n",
    "\n",
    "df = (df.loc[df.chrom == 'X']\n",
    "          .groupby(['chrom', 'species', 'start', 'end'])\n",
    "          .aggregate(numpy.mean)\n",
    "         )    \n",
    "df = df.reset_index()\n",
    "\n",
    "#max_pi_fraction_in_low_regions_apes = 0.2\n",
    "#min_fraction_analyzed_apes = 30000\n",
    "\n",
    "#low_ape_pi_regions = call_low_ape_pi_regions(df, max_pi_fraction_in_low_regions_apes, min_fraction_analyzed_apes)\n",
    "\n",
    "max_pi_in_low_regions_apes = 0.0002\n",
    "min_fraction_analyzed_apes = 30000\n",
    "\n",
    "#low_ape_pi_regions = call_low_ape_pi_regions(df, max_pi_in_low_regions_apes, min_fraction_analyzed_apes)\n",
    "low_ape_pi_regions = call_low_ape_pi_regions(df, min_fraction_analyzed_apes)\n",
    "\n",
    "low_ape_pi_regions.to_hdf(results_dir / 'low_ape_pi_regions.hdf', 'df', mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>species</th>\n",
       "      <th>test</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>run</th>\n",
       "      <th>pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>2</td>\n",
       "      <td>4300000</td>\n",
       "      <td>4500000</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>4</td>\n",
       "      <td>8800000</td>\n",
       "      <td>8900000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>6</td>\n",
       "      <td>9000000</td>\n",
       "      <td>9100000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>8</td>\n",
       "      <td>10900000</td>\n",
       "      <td>11000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>BO</td>\n",
       "      <td>10</td>\n",
       "      <td>16000000</td>\n",
       "      <td>16400000</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom species  test     start       end  run        pi\n",
       "1     X      BO     2   4300000   4500000    2  0.000103\n",
       "3     X      BO     4   8800000   8900000    1  0.000038\n",
       "5     X      BO     6   9000000   9100000    1  0.000112\n",
       "7     X      BO     8  10900000  11000000    1  0.000088\n",
       "9     X      BO    10  16000000  16400000    4  0.000064"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_ape_pi_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BO', 'CC', 'EC', 'ELG', 'NC', 'SO', 'WC', 'WLG'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_ape_pi_regions.species.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute total superset of regions (excl. EWG that was excluded in Nam et al.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.18805883098508822"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genominterv import interval_collapse\n",
    "\n",
    "species_included = ['BO', 'CC', 'EC', 'NC', 'SO', 'WC', 'WLG']#, 'ELG']\n",
    "\n",
    "df = interval_collapse(low_ape_pi_regions.loc[lambda df: df.species.isin(species_included)]\n",
    "                       .sort_values(['chrom', 'start', 'end']))\n",
    "\n",
    "df.assign(length=df.end-df.start).length.sum() / 155270560\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proportions of ILS in 1Mb windows from great ape ILS project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# great_ape_ils_analysis_dir = Path('/home/kmt/projects/great_ape_ils_maps/analyses')\n",
    "great_ape_ils_analysis_dir = data_dir / 'great_ape_ils_maps'\n",
    "\n",
    "window_size = 1000000\n",
    "col_names = [\"start\", \"end\", 'top1', 'top2', 'top3', 'top4']\n",
    "\n",
    "human_chimp_chrX_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/chr1-22_X.statesBin1000000chrX.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_chimp_chrX_ils = (pandas.read_table(human_chimp_chrX_ils_path, names=col_names)\n",
    "                        .assign(chrom = 'X',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "human_orang_chrX_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_ponAbe2_nomLeu1_rheMac2/chr1-22_X.statesBin1000000chrX.import.coalhmm.list_chrHASH-5473120441099014340.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_orang_chrX_ils = (pandas.read_table(human_orang_chrX_ils_path, names=col_names)\n",
    "                        .assign(chrom = 'X',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "\n",
    "human_chimp_chr7_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/chr1-22_X.statesBin1000000chr7.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_chimp_chr7_ils = (pandas.read_table(human_chimp_chr7_ils_path, names=col_names)\n",
    "                        .assign(chrom = '7',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "human_orang_chr7_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_ponAbe2_nomLeu1_rheMac2/chr1-22_X.statesBin1000000chr7.import.coalhmm.list_chrHASH-5473120441099014340.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "human_orang_chr7_ils = (pandas.read_table(human_orang_chr7_ils_path, names=col_names)\n",
    "                        .assign(chrom = '7',\n",
    "                                prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "                                fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "                       )\n",
    "\n",
    "def call_low_ils_regions(df, max_ils, min_analyzed):\n",
    "    df = (df\n",
    "          .assign(prop_ils = lambda df: df.prop_ils.where(df.fraction_analyzed >= min_analyzed)) \n",
    "          .loc[:, ['chrom', 'start', 'end', 'prop_ils']]\n",
    "          .assign(run = 1, \n",
    "                  islow = lambda df: df.prop_ils <= max_ils) # turn masked into nan to break runs at missing data\n",
    "          .assign(test = lambda df: (df.islow != df.islow.shift()).cumsum())\n",
    "          .groupby(['chrom', 'test'])\n",
    "                  .aggregate({'start': 'min', 'end': 'max', 'run': 'sum', 'prop_ils': 'mean'})\n",
    "                  .reset_index()\n",
    "          .loc[lambda df: df.prop_ils <= max_ils, :]\n",
    "         )\n",
    "    return df\n",
    "\n",
    "max_ils_in_low_regions = 0.05\n",
    "minimal_fraction_analyzed_ils = 0.3\n",
    "\n",
    "human_chimp_low_ils_regions_chrX = call_low_ils_regions(human_chimp_chrX_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_chimp_low_ils_regions_chrX.to_hdf(results_dir / 'human_chimp_low_ils_regions_chrX.hdf', 'df', mode='w')\n",
    "\n",
    "human_orang_low_ils_regions_chrX = call_low_ils_regions(human_orang_chrX_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_orang_low_ils_regions_chrX.to_hdf(results_dir / 'human_orang_low_ils_regions_chrX.hdf', 'df', mode='w')\n",
    "\n",
    "\n",
    "human_chimp_low_ils_regions_chr7 = call_low_ils_regions(human_chimp_chr7_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_chimp_low_ils_regions_chr7.to_hdf(results_dir / 'human_chimp_low_ils_regions_chr7.hdf', 'df', mode='w')\n",
    "\n",
    "human_orang_low_ils_regions_chr7 = call_low_ils_regions(human_orang_chr7_ils, \n",
    "                                                        max_ils_in_low_regions, \n",
    "                                                        minimal_fraction_analyzed_ils)\n",
    "human_orang_low_ils_regions_chr7.to_hdf(results_dir / 'human_orang_low_ils_regions_chr7.hdf', 'df', mode='w')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# human_chimp_chr3_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/results/chr1-22_X.statesBin1000000chr3.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "# human_chimp_chr3_ils = (pandas.read_table(human_chimp_chr3_ils_path, names=col_names)\n",
    "#                         .assign(chrom = '3',\n",
    "#                                 prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "#                                 fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "#                        )\n",
    "# df = human_chimp_chr3_ils#.loc[(human_chimp_chr9_ils.start > 80e6) & (human_chimp_chr9_ils.start < 100e6)]\n",
    "# plt.plot(df.start, df.prop_ils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# human_chimp_chr9_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_panTro3_gorGor3_ponAbe2/results/chr1-22_X.statesBin1000000chr9.import.coalhmm.list_chrHASH4563609074768454572.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "# human_chimp_chr9_ils = (pandas.read_table(human_chimp_chr9_ils_path, names=col_names)\n",
    "#                         .assign(chrom = '9',\n",
    "#                                 prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "#                                 fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "#                        )\n",
    "# df = human_chimp_chr9_ils#.loc[(human_chimp_chr9_ils.start > 80e6) & (human_chimp_chr9_ils.start < 100e6)]\n",
    "# plt.plot(df.start, df.prop_ils) ;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# human_orang_chr9_ils_path = great_ape_ils_analysis_dir / \"multiz_hg19_ponAbe2_nomLeu1_rheMac2/results/chr1-22_X.statesBin1000000chr9.import.coalhmm.list_chrHASH-5473120441099014340.hdfmerge.minmaf500_maxgap100_maxn02_junkNn.tbl\"\n",
    "# human_orang_chr9_ils = (pandas.read_table(human_orang_chr9_ils_path, names=col_names)\n",
    "#                         .assign(chrom = '9',\n",
    "#                                 prop_ils = lambda df: (df.top3 + df.top4) / (df.top1 + df.top2 + df.top3 + df.top4),\n",
    "#                                 fraction_analyzed = lambda df: (df.top1 + df.top2 + df.top3 + df.top4) / window_size)\n",
    "#                        )\n",
    "# df = human_orang_chr9_ils.loc[(human_chimp_chr9_ils.start > 80e6) & (human_chimp_chr9_ils.start < 110e6)]\n",
    "# plt.plot(df.start, df.prop_ils) ; "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = call_low_ils_regions(human_orang_chrX_ils, \n",
    "#                            max_ils_in_low_regions, \n",
    "#                            minimal_fraction_analyzed_ils)\n",
    "\n",
    "# df#.loc[(df.start > 50e6) & (df.end < 60e6)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1333229928153677"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_chimp_chrX_ils.loc[human_chimp_chrX_ils.fraction_analyzed >= minimal_fraction_analyzed_ils].prop_ils.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.11592667663464343"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = (human_chimp_chrX_ils\n",
    "      .loc[human_chimp_chrX_ils.fraction_analyzed >= minimal_fraction_analyzed_ils]\n",
    "      .loc[human_chimp_chrX_ils.prop_ils < 0.02]\n",
    "     )\n",
    "len(df) / 155.270560"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Trine Line's circRNA results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "trine_line_genes = pandas.read_csv(data_dir / 'trine_line_data/AnnoTable.csv')\n",
    "trine_line_genes.drop(trine_line_genes.columns[[0]], axis=1, inplace=True)\n",
    "trine_line_x_genes = trine_line_genes.loc[trine_line_genes.chr == 'chrX']\n",
    "trine_line_x_genes.to_hdf(results_dir / 'trine_line_x_genes.hdf', 'df', format='table', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chalmel genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read gene table from Chalmel paper SOM (I exported csv from the excel table):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Description</th>\n",
       "      <th>biological_process terms</th>\n",
       "      <th>molecular_function terms</th>\n",
       "      <th>cellular_component terms</th>\n",
       "      <th>Human_Phenotype</th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Expression in Testis</th>\n",
       "      <th>TF</th>\n",
       "      <th>...</th>\n",
       "      <th>tonsil-non_germinal_center_cells</th>\n",
       "      <th>tonsil-squamous_epithelial_cells</th>\n",
       "      <th>urinary_bladder-urothelial_cells</th>\n",
       "      <th>uterus_post_menopause-cells_in_endometrial_stroma</th>\n",
       "      <th>uterus_post_menopause-glandular_cells</th>\n",
       "      <th>uterus_pre_menopause-cells_in_endometrial_stroma</th>\n",
       "      <th>uterus_pre_menopause-glandular_cells</th>\n",
       "      <th>vagina-squamous_epithelial_cells</th>\n",
       "      <th>vulva_anal_skin-epidermal_cells</th>\n",
       "      <th>Unnamed: 180</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>210262_at</td>\n",
       "      <td>CRISP2</td>\n",
       "      <td>cysteine-rich secretory protein 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>extracellular space (0005615)</td>\n",
       "      <td>-</td>\n",
       "      <td>12</td>\n",
       "      <td>PEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>206112_at</td>\n",
       "      <td>ANKRD7</td>\n",
       "      <td>ankyrin repeat domain 7</td>\n",
       "      <td>male gonad development (0008584)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>11</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1558262_at</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>207116_s_at</td>\n",
       "      <td>GAPDHS</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase, sper...</td>\n",
       "      <td>gluconeogenesis (0006094);glycolysis (0006096)...</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase (NAD+...</td>\n",
       "      <td>cytosol (0005829);microtubule-based flagellum ...</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>231612_at</td>\n",
       "      <td>C4orf35</td>\n",
       "      <td>chromosome 4 open reading frame 35</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  181 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID GeneName                                        Description  \\\n",
       "0    210262_at   CRISP2                  cysteine-rich secretory protein 2   \n",
       "1    206112_at   ANKRD7                            ankyrin repeat domain 7   \n",
       "2   1558262_at        -                                                  -   \n",
       "3  207116_s_at   GAPDHS  glyceraldehyde-3-phosphate dehydrogenase, sper...   \n",
       "4    231612_at  C4orf35                 chromosome 4 open reading frame 35   \n",
       "\n",
       "                            biological_process terms  \\\n",
       "0                                                  -   \n",
       "1                   male gonad development (0008584)   \n",
       "2                                                  -   \n",
       "3  gluconeogenesis (0006094);glycolysis (0006096)...   \n",
       "4                                                  -   \n",
       "\n",
       "                            molecular_function terms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2                                                  -   \n",
       "3  glyceraldehyde-3-phosphate dehydrogenase (NAD+...   \n",
       "4                                                  -   \n",
       "\n",
       "                            cellular_component terms Human_Phenotype Pattern  \\\n",
       "0                      extracellular space (0005615)               -      12   \n",
       "1                                                  -               -      11   \n",
       "2                                                  -               -      13   \n",
       "3  cytosol (0005829);microtubule-based flagellum ...               -      13   \n",
       "4                                                  -               -      13   \n",
       "\n",
       "  Expression in Testis TF  ... tonsil-non_germinal_center_cells  \\\n",
       "0                PEHET  -  ...                                0   \n",
       "1                SEHET  -  ...                                -   \n",
       "2                SEHET  -  ...                                -   \n",
       "3                SEHET  -  ...                                -   \n",
       "4                SEHET  -  ...                                -   \n",
       "\n",
       "  tonsil-squamous_epithelial_cells urinary_bladder-urothelial_cells  \\\n",
       "0                                0                                0   \n",
       "1                                -                                -   \n",
       "2                                -                                -   \n",
       "3                                -                                -   \n",
       "4                                -                                -   \n",
       "\n",
       "  uterus_post_menopause-cells_in_endometrial_stroma  \\\n",
       "0                                                 0   \n",
       "1                                                 -   \n",
       "2                                                 -   \n",
       "3                                                 -   \n",
       "4                                                 -   \n",
       "\n",
       "   uterus_post_menopause-glandular_cells  \\\n",
       "0                                      0   \n",
       "1                                      -   \n",
       "2                                      -   \n",
       "3                                      -   \n",
       "4                                      -   \n",
       "\n",
       "   uterus_pre_menopause-cells_in_endometrial_stroma  \\\n",
       "0                                                 0   \n",
       "1                                                 -   \n",
       "2                                                 -   \n",
       "3                                                 -   \n",
       "4                                                 -   \n",
       "\n",
       "   uterus_pre_menopause-glandular_cells  vagina-squamous_epithelial_cells  \\\n",
       "0                                     0                                 0   \n",
       "1                                     -                                 -   \n",
       "2                                     -                                 -   \n",
       "3                                     -                                 -   \n",
       "4                                     -                                 -   \n",
       "\n",
       "   vulva_anal_skin-epidermal_cells  Unnamed: 180  \n",
       "0                                0           NaN  \n",
       "1                                -           NaN  \n",
       "2                                -           NaN  \n",
       "3                                -           NaN  \n",
       "4                                -           NaN  \n",
       "\n",
       "[5 rows x 181 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chalmel_genes = pandas.read_csv(data_dir / 'chalmel_genes.csv', sep=';')\n",
    "chalmel_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Description</th>\n",
       "      <th>biological_process terms</th>\n",
       "      <th>molecular_function terms</th>\n",
       "      <th>cellular_component terms</th>\n",
       "      <th>Human_Phenotype</th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Expression in Testis</th>\n",
       "      <th>TF</th>\n",
       "      <th>...</th>\n",
       "      <th>vagina-squamous_epithelial_cells</th>\n",
       "      <th>vulva_anal_skin-epidermal_cells</th>\n",
       "      <th>Unnamed: 180</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>210262_at</td>\n",
       "      <td>CRISP2</td>\n",
       "      <td>cysteine-rich secretory protein 2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>extracellular space (0005615)</td>\n",
       "      <td>-</td>\n",
       "      <td>12</td>\n",
       "      <td>PEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000124490</td>\n",
       "      <td>6</td>\n",
       "      <td>49660073</td>\n",
       "      <td>49681274</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CRISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>206112_at</td>\n",
       "      <td>ANKRD7</td>\n",
       "      <td>ankyrin repeat domain 7</td>\n",
       "      <td>male gonad development (0008584)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>11</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000106013</td>\n",
       "      <td>7</td>\n",
       "      <td>117854727</td>\n",
       "      <td>117882785</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ANKRD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>207116_s_at</td>\n",
       "      <td>GAPDHS</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase, sper...</td>\n",
       "      <td>gluconeogenesis (0006094);glycolysis (0006096)...</td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase (NAD+...</td>\n",
       "      <td>cytosol (0005829);microtubule-based flagellum ...</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000105679</td>\n",
       "      <td>19</td>\n",
       "      <td>36024314</td>\n",
       "      <td>36036218</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GAPDHS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>243208_x_at</td>\n",
       "      <td>ACTL9</td>\n",
       "      <td>actin-like 9</td>\n",
       "      <td>-</td>\n",
       "      <td>protein binding (0005515)</td>\n",
       "      <td>cytoplasm (0005737);cytoskeleton (0005856)</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000263000</td>\n",
       "      <td>HG729_PATCH</td>\n",
       "      <td>8762287</td>\n",
       "      <td>8763706</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACTL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>243208_x_at</td>\n",
       "      <td>ACTL9</td>\n",
       "      <td>actin-like 9</td>\n",
       "      <td>-</td>\n",
       "      <td>protein binding (0005515)</td>\n",
       "      <td>cytoplasm (0005737);cytoskeleton (0005856)</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>SEHET</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000181786</td>\n",
       "      <td>19</td>\n",
       "      <td>8807751</td>\n",
       "      <td>8809172</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ACTL9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID GeneName                                        Description  \\\n",
       "0    210262_at   CRISP2                  cysteine-rich secretory protein 2   \n",
       "1    206112_at   ANKRD7                            ankyrin repeat domain 7   \n",
       "2  207116_s_at   GAPDHS  glyceraldehyde-3-phosphate dehydrogenase, sper...   \n",
       "3  243208_x_at    ACTL9                                       actin-like 9   \n",
       "4  243208_x_at    ACTL9                                       actin-like 9   \n",
       "\n",
       "                            biological_process terms  \\\n",
       "0                                                  -   \n",
       "1                   male gonad development (0008584)   \n",
       "2  gluconeogenesis (0006094);glycolysis (0006096)...   \n",
       "3                                                  -   \n",
       "4                                                  -   \n",
       "\n",
       "                            molecular_function terms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2  glyceraldehyde-3-phosphate dehydrogenase (NAD+...   \n",
       "3                          protein binding (0005515)   \n",
       "4                          protein binding (0005515)   \n",
       "\n",
       "                            cellular_component terms Human_Phenotype Pattern  \\\n",
       "0                      extracellular space (0005615)               -      12   \n",
       "1                                                  -               -      11   \n",
       "2  cytosol (0005829);microtubule-based flagellum ...               -      13   \n",
       "3         cytoplasm (0005737);cytoskeleton (0005856)               -      13   \n",
       "4         cytoplasm (0005737);cytoskeleton (0005856)               -      13   \n",
       "\n",
       "  Expression in Testis TF  ... vagina-squamous_epithelial_cells  \\\n",
       "0                PEHET  -  ...                                0   \n",
       "1                SEHET  -  ...                                -   \n",
       "2                SEHET  -  ...                                -   \n",
       "3                SEHET  -  ...                                2   \n",
       "4                SEHET  -  ...                                2   \n",
       "\n",
       "  vulva_anal_skin-epidermal_cells Unnamed: 180   Gene stable ID        chrom  \\\n",
       "0                               0          NaN  ENSG00000124490            6   \n",
       "1                               -          NaN  ENSG00000106013            7   \n",
       "2                               -          NaN  ENSG00000105679           19   \n",
       "3                               2          NaN  ENSG00000263000  HG729_PATCH   \n",
       "4                               2          NaN  ENSG00000181786           19   \n",
       "\n",
       "       start        end  strand       Gene type    name  \n",
       "0   49660073   49681274      -1  protein_coding  CRISP2  \n",
       "1  117854727  117882785       1  protein_coding  ANKRD7  \n",
       "2   36024314   36036218       1  protein_coding  GAPDHS  \n",
       "3    8762287    8763706      -1  protein_coding   ACTL9  \n",
       "4    8807751    8809172      -1  protein_coding   ACTL9  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chalmel_genes = chalmel_genes.merge(biomart_genes, left_on=['GeneName'], right_on=['name'], how='inner')\n",
    "chalmel_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chalmel_genes.loc[(chalmel_genes.chrom == '3') & (chalmel_genes.start > 45e6) & (chalmel_genes.end < 55e6)].sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "chalmel_genes.to_hdf(results_dir / 'chalmel_genes.hdf', 'df', format='table', mode='w')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
